### PR TITLE
Simplify daml assistant interface.

### PR DESCRIPTION
**Pull request by associahedron**
Reorganizing the command-line interface to make it more intuitive.

In particular:

* Hid some low-level commands (`daml exec` and `daml ide`).
* Introduced the `daml damlc` command which just invokes the appropriately versioned `damlc`. 
* Removed `daml init` and `daml package` since they are subsumed by `daml build` and the latter command does not require users to manage the state of package database. These commands are still available as `daml damlc init` and `daml damlc new-package`, but not encouraged of course.
* Removed `daml test` for now, since it isn't DAML project aware. I will create an issue to add a `daml test` command that understands `daml.yaml` and finds the scenarios to run automatically (and fails when run outside of a project). The old `daml test` command can still be invoked by `daml damlc test`.
* Tried to reorder and describe the commands in an intuitive fashion, so someone can figure out which commands are meant to be run in a project, which commands deal with SDK versions, which commands just launch other programs, etc.

### CODE REVIEW

**Comment by neil-da** [comment]
Can you paste the output of `--help`, since it seems that's what you are really trying to change in this PR.

**Comment by associahedron** [comment]
> Can you paste the output of `--help`, since it seems that's what you are really trying to change in this PR.

Here it is:
```
$ daml-head --help
Usage: daml COMMAND

Available options:
  -h,--help                Show this help text

Available commands:
  version                  Display DAML SDK version
  install                  Install DAML SDK version
  studio                   Launch DAML studio
  new                      Create a new project
  build                    Build project
  start                    Run sandbox and navigator in project
  damlc                    Launch the DAML compiler
  sandbox                  Launch the sandbox
  navigator                Launch the navigator
  extractor                Launch the extractor
```

File: daml-assistant/src/DAML/Assistant/Command.hs [hunk 1]
```diff
@@ -46,13 +46,13 @@ dispatch info = subcommand
 commandParser :: [SdkCommandInfo] -> Parser Command
 commandParser cmds | (hidden, visible) <- partition isHidden cmds = asum
     [ subparser -- visible commands
-        $  builtin "version" "Display SDK version" mempty (pure Version <**> helper)
-        <> builtin "install" "Install SDK version" mempty (Install <$> installParser <**> helper)
-        <> builtin "exec" "Execute command with daml environment." forwardOptions
-            (Exec <$> strArgument (metavar "CMD") <*> many (strArgument (metavar "ARGS")))
+        $  builtin "version" "Display DAML SDK version" mempty (pure Version <**> helper)
+        <> builtin "install" "Install DAML SDK version" mempty (Install <$> installParser <**> helper)
```
Comment (bethaitman) [inline daml-assistant/src/DAML/Assistant/Command.hs]:
```suggestion
        <> builtin "install" "Install the specified DAML SDK version, or the latest version if no version is specified" mempty (Install <$> installParser <**> helper)
```

File: daml-assistant/src/DAML/Assistant/Command.hs [hunk 2]
```diff
@@ -46,13 +46,13 @@ dispatch info = subcommand
 commandParser :: [SdkCommandInfo] -> Parser Command
 commandParser cmds | (hidden, visible) <- partition isHidden cmds = asum
     [ subparser -- visible commands
-        $  builtin "version" "Display SDK version" mempty (pure Version <**> helper)
-        <> builtin "install" "Install SDK version" mempty (Install <$> installParser <**> helper)
-        <> builtin "exec" "Execute command with daml environment." forwardOptions
-            (Exec <$> strArgument (metavar "CMD") <*> many (strArgument (metavar "ARGS")))
+        $  builtin "version" "Display DAML SDK version" mempty (pure Version <**> helper)
```
Comment (bethaitman) [inline daml-assistant/src/DAML/Assistant/Command.hs]: For what? Helps to be specific - I guess this is something like "Display all installed DAML SDK versions. If run inside a project, also shows the SDK version that the project is using"

File: release/sdk-config.yaml.tmpl [hunk 3]
```diff
@@ -1,50 +1,36 @@
...
-  path: damlc/da-hs-damlc-app
-  args: ["test"]
-  desc: "Run the scenarios in the given DAML file and all dependencies"
-- name: ide
-  path: damlc/da-hs-damlc-app
-  args: ["lax", "ide"]
-  desc: "Run the DAML language server"
 - name: studio
   path: daml-helper/daml-helper
   desc: "Launch DAML studio"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Launch DAML Studio"
```

File: release/sdk-config.yaml.tmpl [hunk 4]
```diff
@@ -1,50 +1,36 @@
...
+  path: daml-helper/daml-helper
+  args: ["start"]
+  desc: "Run sandbox and navigator in project"
+- name: damlc
+  path: damlc/da-hs-damlc-app
+  desc: "Launch the DAML compiler"
 - name: sandbox
   path: daml-helper/daml-helper
-  desc: "Start the sandbox"
+  desc: "Launch the sandbox"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Launch the Sandbox"
```

File: release/sdk-config.yaml.tmpl [hunk 5]
```diff
@@ -1,50 +1,36 @@
...
   args: ["run-jar", "sandbox/sandbox.jar"]
 - name: navigator
   path: daml-helper/daml-helper
-  desc: "Start the navigator"
+  desc: "Launch the navigator"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Launch the Navigator"
```

File: release/sdk-config.yaml.tmpl [hunk 6]
```diff
@@ -1,50 +1,36 @@
...
   args: ["run-jar", "navigator/navigator.jar"]
 - name: extractor
   path: daml-helper/daml-helper
-  desc: "Start the extractor"
+  desc: "Launch the extractor"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Launch the Extractor"
```

File: release/sdk-config.yaml.tmpl [hunk 7]
```diff
@@ -1,50 +1,36 @@
...
+  desc: "Create a new project"
+  args: ["new"]
+- name: build
+  path: damlc/da-hs-damlc-app
+  args: ["build"]
+  desc: "Build project"
+- name: start
+  path: daml-helper/daml-helper
+  args: ["start"]
+  desc: "Run sandbox and navigator in project"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Launch the Sandbox and the Navigator for the current DAML project"
```

File: release/sdk-config.yaml.tmpl [hunk 8]
```diff
@@ -1,50 +1,36 @@
...
-  path: damlc/da-hs-damlc-app
-  args: ["lax", "ide"]
-  desc: "Run the DAML language server"
 - name: studio
   path: daml-helper/daml-helper
   desc: "Launch DAML studio"
   args: ["studio"]
+- name: new
+  path: daml-helper/daml-helper
+  desc: "Create a new project"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Create a new DAML project"
```

File: release/sdk-config.yaml.tmpl [hunk 9]
```diff
@@ -1,50 +1,36 @@
...
+  args: ["new"]
+- name: build
+  path: damlc/da-hs-damlc-app
+  args: ["build"]
+  desc: "Build project"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]:
```suggestion
  desc: "Build the current DAML project into a DAR file"
```

File: release/sdk-config.yaml.tmpl [hunk 10]
```diff
@@ -1,50 +1,36 @@
...
+- name: start
+  path: daml-helper/daml-helper
+  args: ["start"]
+  desc: "Run sandbox and navigator in project"
+- name: damlc
+  path: damlc/da-hs-damlc-app
+  desc: "Launch the DAML compiler"
```
Comment (bethaitman) [inline release/sdk-config.yaml.tmpl]: Surely the compiler is something you run? I guess launch to me implies an ongoing process.

File: daml-assistant/src/DAML/Assistant/Command.hs (hunk 1 above)
Comment (neil-da) [reply daml-assistant/src/DAML/Assistant/Command.hs]: `Install DAML SDK version (defaults to latest)` seems to capture the same thing more concisely (given how `--help` shows, concise matters)

File: release/sdk-config.yaml.tmpl (hunk 10 above)
Comment (neil-da) [reply release/sdk-config.yaml.tmpl]: Agreed. Run

**Comment by neil-da** [comment]
We use the word SDK in a few purposes. Does it mean anything? Can it be omitted?

**Comment by bethaitman** [comment]
SDK is the thing you install. I think we only have to use it when we're referring to that thing (ie installing, upgrading, or uninstalling it)

File: daml-assistant/src/DAML/Assistant/Command.hs (hunk 1 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Command.hs]: It does not default to latest (that is the command `daml install latest`, with `daml install` showing all the available targets). I like the description "Install the specified DAML SDK version".

File: daml-assistant/src/DAML/Assistant/Command.hs (hunk 2 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Command.hs]: It really only displays the SDK version in use in this specific location. (e.g. inside a project it displays the project's SDK version, outside the project it displays the latest version). It's probably not a very useful command, would be better as an option `--version` that shows the current SDK and DAML assistant version.

File: release/sdk-config.yaml.tmpl (hunk 8 above)
Comment (associahedron) [reply release/sdk-config.yaml.tmpl]: :+1:

File: release/sdk-config.yaml.tmpl (hunk 3 above)
Comment (associahedron) [reply release/sdk-config.yaml.tmpl]: :+1:

File: release/sdk-config.yaml.tmpl (hunk 7 above)
Comment (associahedron) [reply release/sdk-config.yaml.tmpl]: This sadly becomes split into two lines. Is there a shorter way to say this?

Comment (bethaitman) [reply release/sdk-config.yaml.tmpl]: `"Launch Sandbox and Navigator for the current DAML project"? Can even remove the other "the" if we have to

Comment (associahedron) [reply release/sdk-config.yaml.tmpl]: Thanks :-) removing all the 'the's was the trick.

File: daml-assistant/integration-tests/run-test.sh [hunk 11]
```diff
@@ -1,21 +0,0 @@
-#!/usr/bin/env bash
```
Comment (cocreature) [inline daml-assistant/integration-tests/run-test.sh]: Good catch :smile:

File: daml-assistant/integration-tests/src/Main.hs [hunk 12]
```diff
@@ -59,15 +59,10 @@ quickstartTests :: FilePath -> FilePath -> TestTree
...
-    , testCase "daml init" $ withCurrentDirectory quickstartDir $
-          callProcessQuiet "daml" ["init"]
-    , testCase "daml package" $ withCurrentDirectory quickstartDir $
-          -- This location is assumed by the codegen in the quickstart example.
-          callProcessQuiet "daml" ["package", "-o", "target/daml/iou.dar"]
     , testCase "daml build " $ withCurrentDirectory quickstartDir $
           callProcessQuiet "daml" ["build"]
-    , testCase "daml test" $ withCurrentDirectory quickstartDir $
-          callProcessQuiet "daml" ["test", "daml/Main.daml"]
+    , testCase "daml damlc test" $ withCurrentDirectory quickstartDir $
```
Comment (cocreature) [inline daml-assistant/integration-tests/src/Main.hs]: I opened https://github.com/digital-asset/daml/issues/743 to track the changes to `damlc test` before we add `daml test` back.

File: daml-assistant/integration-tests/src/Main.hs [hunk 13]
```diff
@@ -59,15 +59,10 @@ quickstartTests :: FilePath -> FilePath -> TestTree
 quickstartTests quickstartDir mvnDir = testGroup "quickstart"
     [ testCase "daml new" $
           callProcessQuiet "daml" ["new", quickstartDir]
-    , testCase "daml init" $ withCurrentDirectory quickstartDir $
-          callProcessQuiet "daml" ["init"]
-    , testCase "daml package" $ withCurrentDirectory quickstartDir $
-          -- This location is assumed by the codegen in the quickstart example.
-          callProcessQuiet "daml" ["package", "-o", "target/daml/iou.dar"]
```
Comment (cocreature) [inline daml-assistant/integration-tests/src/Main.hs]: Removing this breaks the `mvn compile` test below which assumes this location. You could just add the `-o` parameter to `daml build` instead.

File: daml-assistant/integration-tests/src/Main.hs (hunk 12 above)
Comment (associahedron) [reply daml-assistant/integration-tests/src/Main.hs]: Thanks!

File: daml-assistant/integration-tests/src/Main.hs (hunk 13 above)
Comment (associahedron) [reply daml-assistant/integration-tests/src/Main.hs]: Ah, thanks!

**Pull request merged**